pageserver/src/tenant/timeline.rs

```diff
@@ -2680,52 +2680,57 @@ impl Timeline {
+        //
+        // Some unit tests depend on garbage-collection working even when
+        // CLOG data is missing, so that find_lsn_for_timestamp() doesn't
+        // work, so avoid calling it altogether if time-based retention is not
+        // configured. It would be pointless anyway.
+        let pitr_cutoff = if pitr != Duration::ZERO {
             let now = SystemTime::now();
             if let Some(pitr_cutoff_timestamp) = now.checked_sub(pitr) {
                 let pitr_timestamp = to_pg_timestamp(pitr_cutoff_timestamp);
 
-                match self
-                    .find_lsn_for_timestamp(pitr_timestamp)
-                    .no_ondemand_download()?
+                match with_ondemand_download(|| self.find_lsn_for_timestamp(pitr_timestamp)).await?
                 {
-                    LsnForTimestamp::Present(lsn) => pitr_cutoff_lsn = lsn,
+                    LsnForTimestamp::Present(lsn) => lsn,
                     LsnForTimestamp::Future(lsn) => {
+                        // The timestamp is in the future. That sounds impossible,
+                        // but what it really means is that there hasn't been
+                        // any commits since the cutoff timestamp.
                         debug!("future({})", lsn);
-                        pitr_cutoff_lsn = gc_info.horizon_cutoff;
+                        cutoff_horizon
                     }
                     LsnForTimestamp::Past(lsn) => {
                         debug!("past({})", lsn);
+                        // conservative, safe default is to remove nothing, when we
+                        // have no commit timestamp data available
+                        *self.get_latest_gc_cutoff_lsn()
                     }
                     LsnForTimestamp::NoData(lsn) => {
                         debug!("nodata({})", lsn);
+                        // conservative, safe default is to remove nothing, when we
+                        // have no commit timestamp data available
+                        *self.get_latest_gc_cutoff_lsn()
                     }
                 }
-                debug!("pitr_cutoff_lsn = {:?}", pitr_cutoff_lsn)
+            } else {
+                // If we don't have enough data to convert to LSN,
+                // play safe and don't remove any layers.
+                *self.get_latest_gc_cutoff_lsn()
             }
         } else {
-            // No time-based retention. (Some unit tests depend on garbage-collection
-            // working even when CLOG data is missing, so that find_lsn_for_timestamp()
-            // above doesn't work.)
-            pitr_cutoff_lsn = gc_info.horizon_cutoff;
-        }
-        gc_info.pitr_cutoff = pitr_cutoff_lsn;
+            // No time-based retention was configured. Set time-based cutoff to
+            // same as LSN based.
+            cutoff_horizon
+        };
+
+        // Grab the lock and update the values
+        *self.gc_info.write().unwrap() = GcInfo {
```

Comment:
I wonder, if races are possible now?
`update_gc_info` gets called (via `refresh_gc_info_internal`) from `tenant::gather_size_inputs`.

Can there happen a situation when first GC will be downloading the layers, then some HTTP request comes on all layers downloaded and updates GC info, then the GC call would overwrite that GC info with something older?

Reply:
`refresh_gc_info_internal` is the only direct caller of `update_gc_info`, and it holds the `gc_cs` lock. So no, there can be only one caller of `update_gc_info` at a time. I added a comment on that in `update_gc_info`.

Without that, this would indeed be racy. With the LSN and time-based cutoff values, I think that'd be OK. The worst that could happen is that the cutoff value moves backwards, but that wouldn't cause any trouble on its own. When GC actually removes something, it uses the `latest_gc_cutoff_lsn` value. These values in the gc_info determine what `latest_gc_cutoff_lsn` can be advanced to, but nothing breaks AFAICS if they move back and forth.

`retain_lsns` is more tricky. If a new branch is created, it's important that the branch point's LSN is included in `retain_lsns`. The `gc_cs` lock ensures that.